include/quadrilateral_gimp_element.tcc

```diff
@@ -0,0 +1,250 @@
+//!   13----------12----------11----------10
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |        (-1, 1)      (1,1)         |
+//!   14----------3-----------2-----------9
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |particle   |           |
+//!   |           | location  |           |
+//!   |           |           |           |
+//!   15----------0-----------1-----------8
+//!   |        (-1,-1)      (1,-1)        |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   4-----------5-----------6-----------7
+
+// RENAME TO QUAD GIMP
+//! Return shape functions of a 16-node Quadrilateral GIMP Element at a given
+//! local coordinate
+template <>
+inline Eigen::VectorXd
+    mpm::QuadrilateralGIMPElement<2, 16>::QuadrilateralElement<2, 4>::shapefn(
+        const Eigen::Matrix<double, 2, 1>& xi, const VectorDim& particle_size,
+        const Eigen::Matrix<double, 2, 1>& deformation_gradient) const {
+
+  Eigen::Matrix<double, 16, 1> shapefn;
+
+  //! Matrix to store local node coordinates
+  Eigen::Matrix<double, 16, 2> local_node;
+  // clang-format off
+  local_node << -1., -1.,
+                 1., -1.,
+                -1.,  1.,
+                 1.,  1.,
+                -3., -3.,
+                -1., -3.,
+                 1., -3.,
+                 3., -3.,
+                 3., -1.,
+                 3.,  1.,
+                 3.,  3.,
+                 1.,  3.,
+                -1.,  3.,
+                -3.,  3.,
+                -3.,  1.,
+                -3., -1.;
+  // clang-format on
+
+  //! length and volume of element in local coordinate (TODO: double check)
+  const double element_length = 2;
+  const double element_volume = 4;  // 2*2
+
+  //! local node x & y valyes
+  double xni, yni;
+  const double numnodes = 16;
+
+  //! local particle - local node
+  double xpxi, ypyi;
+
+  //! local shapefunctions for x and y coordinates
+  double sxni, syni;
+
+  for (unsigned n = 0; n < numnodes; n++) {
+    xni = local_node(n, 0);
```

Comment:
Refer uintah shape functions for GIMP to see how the do it without having tod define local coordinates